src/main/java/networkservlet/NetworkServlet.java

```diff
@@ -95,59 +92,73 @@ protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws Se
+                        ref = new SimpleBranch.Builder().displayId(displayId).id(id).latestCommit(hash).build();
+                    }
+                    labels.get(hash).add(ref);
+                }
+            }
+        }
         return labels;
     }
 
     protected Page<Commit> getCommits(final Repository repository,
-                                            final Map<String, List<Ref>> labels,
                                             final Integer limit,
                                             final Integer offset) {
-        final ArrayList<Commit> commits = new ArrayList<>();
-        TraversalRequest request = new TraversalRequest.Builder().repository(repository).include(labels.keySet()).build();
-        final ApplicationUser user = authenticationContext.getCurrentUser();
-        commitService.traverse(request, new TraversalCallback() {
-            private Integer counter;
-            @Override
-            public void onStart(@Nonnull TraversalContext context) {
-                                                        this.counter = 0;
-                                                                         }
-            @Override
-            public TraversalStatus onNode(final @Nonnull CommitGraphNode node) {
-                if (counter >= offset && counter < (offset + limit)) {
-                    securityService.impersonating(user, "Reading repository commits")
-                        .call(() -> {
-                            commits.add(commitService.getCommit(
-                                    new CommitRequest.Builder(repository, node.getCommit().getId()).build()));
-                            return true;
-                        });
-                } else if (counter >= (offset + limit)) {
-                    return TraversalStatus.FINISH;
-                }
-                // If there are no parents then we are at the root node.
-                if (node.getParents().size() > 0) {
-                    counter++;
-                    return TraversalStatus.CONTINUE;
-                } else {
-                    return TraversalStatus.FINISH;
-                }
-            }
-        });
-
-        // Convert the arraylist to a page
-        return PageUtils.createPage(commits, PageUtils.newRequest(0, commits.size()));
+        PageRequest pr = new PageRequestImpl( offset, limit );
+        pr = pr.buildRestrictedPageRequest( 50 );
+        CommitsCommandParameters ccp = new CommitsCommandParameters.Builder().all( true ).build();
+        PagedCommitOutputHandler pcoh = new PagedCommitOutputHandler( repository, ccp, pr );
+        GitScmCommandBuilder revList = gitScm.getCommandBuilderFactory().builder(repository).command("rev-list");
+        revList.argumentAt(0, "--pretty=" + FORMAT );
+        revList.argumentAt(1, "--branches=*" );
+        revList.argumentAt(2, "--tags=*" );
+        revList.argumentAt(3, "--topo-order" );
```

Comment:
I believe this truly fixes #24 with regards to out of order commits. The issue being Stash didn't ensure the second clause of `--topo-order`

> avoid showing commits on multiple lines of history intermixed.